modules/client/src/main/java/org/apache/ignite/internal/client/table/ClientKeyValueView.java

```diff
@@ -544,6 +546,27 @@ public CompletableFuture<Void> streamData(Publisher<Entry<K, V>> publisher, @Nul
         return ClientDataStreamer.streamData(publisher, opts, batchSender, provider, tbl);
     }
 
+    /** {@inheritDoc} */
+    @Override
+    protected @Nullable Function<SqlRow, Entry<K, V>> queryMapper(ResultSetMetadata meta, ClientSchema schema) {
+        int[] keyMapping = indexMapping(schema.columns(), 0, schema.keyColumnCount(), meta);
+        int[] valMapping = indexMapping(schema.columns(), schema.keyColumnCount(), schema.columns().length, meta);
+
+        Marshaller keyMarsh = schema.getMarshaller(keySer.mapper(), TuplePart.KEY, true);
+        Marshaller valMarsh = schema.getMarshaller(valSer.mapper(), TuplePart.VAL, true);
+
+        return (row) -> {
+            try {
+                return new IgniteBiTuple<>(
+                        (K) keyMarsh.readObject(new TupleReader(new SqlRowProjection(row, keyMapping)), null),
+                        (V) valMarsh.readObject(new TupleReader(new SqlRowProjection(row, valMapping)), null)
+                );
+            } catch (MarshallerException e) {
+                throw new IgniteException(INTERNAL_ERR, "Failed to map query results: " + e.getMessage(), e);
```

Comment:
Should we rethrow as public `org.apache.ignite.lang.MarshallerException` instead?

Reply:
Done